### PR TITLE
language model decode 

### PR DESCRIPTION
**Pull request by 2877992943**
here is two part of  change for inference for language model problem, line 301 and line 390
1)line 301 ,input tensor with shape (?,?,1) obviously don't have 4th dimension, so squeeze axis=[2,3] not work
2)line 390, when setting beam_size >1 ,like beam_size=6,  ret["outputs"] with shape [batch size,beam_size,steps],  the second dimension adjustment for ret["outputs"][:, partial_targets_length:] intend to change the third dimension ,thus should like this ret["outputs"][:, :,partial_targets_length:]

### CODE REVIEW

**Comment by 2877992943** [comment]
I signed it!

**Comment by lukaszkaiser** [comment]
Please remove the .DS_STORE files from this PR and re-run travis to make test pass. Thanks!

File: tensor2tensor/models/transformer.py [hunk 1]
```diff
@@ -45,6 +45,13 @@
 from tensorflow.python.util import nest
 
 
+#changed
```
Comment (lukaszkaiser) [inline tensor2tensor/models/transformer.py]: Please remove the unnecessary comments.

File: tensor2tensor/models/transformer.py [hunk 2]
```diff
@@ -298,9 +306,14 @@ def _fast_decode(self,
       # We force the outputs to begin with these sequences.
       encoder_output = None
       encoder_decoder_attention_bias = None
-      partial_targets = tf.squeeze(tf.to_int64(features["inputs"]), [2, 3])
+      if len(features["inputs"].shape) >= 4:
+          partial_targets = tf.squeeze(tf.to_int64(features["inputs"]), [2, 3])
+      elif len(features["inputs"].shape) < 4:  # (batch,step,1)
+          partial_targets = tf.squeeze(tf.to_int64(features["inputs"]), [2])# partial target-> [?,?]
+      #partial_targets = tf.squeeze(tf.to_int64(features["inputs"]), [2, 3])
```
Comment (lukaszkaiser) [inline tensor2tensor/models/transformer.py]: Here too.

File: tensor2tensor/models/transformer.py [hunk 3]
```diff
@@ -386,8 +399,13 @@ def forced_logits():
         top_beams=top_beams,
         alpha=alpha,
         batch_size=batch_size)
-    if partial_targets is not None:
-      ret["outputs"] = ret["outputs"][:, partial_targets_length:]
+    #if partial_targets is not None:
```
Comment (lukaszkaiser) [inline tensor2tensor/models/transformer.py]: And here.

**Comment by lukaszkaiser** [comment]
Thanks! Just a few nits and need to sync to current state, looks good otherwise.

File: tensor2tensor/models/transformer.py [hunk 4]
```diff
@@ -386,8 +399,13 @@ def forced_logits():
...
+    if partial_targets is not None and beam_size==1:
+      ret["outputs"] = ret["outputs"][:, partial_targets_length:] # beam_size=10 [batch,beam,steps]
+    elif partial_targets is not None and beam_size>1:
+        ret["outputs"] = ret["outputs"][:, :,partial_targets_length:]
```
Comment (rasmi) [inline tensor2tensor/models/transformer.py]: Add a space after the comma.

File: tensor2tensor/models/transformer.py [hunk 5]
```diff
@@ -487,6 +505,7 @@ def inner_loop(i, finished, next_id, decoded_ids, cache):
     def is_not_finished(i, finished, *_):
       return (i < decode_length) & tf.logical_not(tf.reduce_all(finished))
 
+
```
Comment (rasmi) [inline tensor2tensor/models/transformer.py]: Remove this newline.

File: tensor2tensor/models/transformer.py [hunk 6]
```diff
@@ -386,8 +399,13 @@ def forced_logits():
...
         alpha=alpha,
         batch_size=batch_size)
-    if partial_targets is not None:
-      ret["outputs"] = ret["outputs"][:, partial_targets_length:]
+    #if partial_targets is not None:
+    if partial_targets is not None and beam_size==1:
+      ret["outputs"] = ret["outputs"][:, partial_targets_length:] # beam_size=10 [batch,beam,steps]
+    elif partial_targets is not None and beam_size>1:
+        ret["outputs"] = ret["outputs"][:, :,partial_targets_length:]
+
```
Comment (rasmi) [inline tensor2tensor/models/transformer.py]: Remove these newlines

File: tensor2tensor/models/transformer.py [hunk 7]
```diff
@@ -298,9 +306,14 @@ def _fast_decode(self,
...
       encoder_decoder_attention_bias = None
-      partial_targets = tf.squeeze(tf.to_int64(features["inputs"]), [2, 3])
+      if len(features["inputs"].shape) >= 4:
+          partial_targets = tf.squeeze(tf.to_int64(features["inputs"]), [2, 3])
+      elif len(features["inputs"].shape) < 4:  # (batch,step,1)
+          partial_targets = tf.squeeze(tf.to_int64(features["inputs"]), [2])# partial target-> [?,?]
+      #partial_targets = tf.squeeze(tf.to_int64(features["inputs"]), [2, 3])
       partial_targets_length = common_layers.shape_list(partial_targets)[1]
       decode_length += partial_targets_length
+      #decode_length=partial_targets_length+features['decode_length']# input_len + extra_len
```
Comment (rasmi) [inline tensor2tensor/models/transformer.py]: Remove comment.

File: tensor2tensor/models/transformer.py (hunk 4 above)
Comment (rasmi) [reply tensor2tensor/models/transformer.py]: Add spaces before and after '==' and '>'.

Comment (rasmi) [reply tensor2tensor/models/transformer.py]: Remove comment.

**Comment by yynil** [comment]
hi,@rasmi that fix doesn't fix the same bugs in Language model decoding. But it's related. This fix can help to generate correct beam search results. The https://github.com/tensorflow/tensor2tensor/pull/652 just fixed another bug that would terminate the decoding because of wrong tensor shape.

**Comment by rasmi** [comment]
Thanks @yynil, that sounds good. Please make the requested changes when you have the time so this change can be merged -- thanks!

**Comment by lukaszkaiser** [comment]
Can you guys sync this with the most recent branch so we can merge? Great thanks for doing this!

**Comment by yynil** [comment]
hi,@lukaszkaiser ,@2877992943  and I are in the same team . I have pushed the changes in this PR https://github.com/tensorflow/tensor2tensor/pull/679 and I think this PR can be closed safely. The modifications are already in 1.5.6.

**Comment by lukaszkaiser** [comment]
Thanks! Closing in that case, but great thanks again and make a new one if stuff breaks again!
